### DISCUSSION TITLE
Severe performance issues with global CockroachDB

### DISCUSSION THREAD

**Discussion by vstreame** [ory/kratos#3134] [Q&A]
Hi there,

We're self-hosting Kratos using CockroachDB as our backing datastore. We chose CockroachDB as we are in 4 datacenters around the globe all running in master-master mode with one another. But ever since introducing Kratos we've seen latency increase on avg by a full second (1000ms) when someone is signed in vs not. Tracking this down it looks like the issues is in CDB. For example this query:

```
SELECT session_devices.created_at,
       session_devices.id,
       session_devices.ip_address,
       session_devices.location,
       session_devices.nid,
       session_devices.session_id,
       session_devices.updated_at,
       session_devices.user_agent
  FROM session_devices AS session_devices
 WHERE session_id = $1
```

Takes a whopping 230ms on average. Not something I would expect from a primary key lookup. Talking to CDB they say it's because we haven't marked our tables as "GLOBAL" so it may have to go to another region often to find the data. Is this something Kratos should be setting via it's migrations? Or should _we_ have known to set this up? Right now this performance isn't going to fly so we either need to rip our Kratos or figure a way to configure out CDB database better. This was unfortunately our first foray into both Kratos and CDB and we were hoping it would "just work" but doesn't seem to be the case 😢 

Any and all help would be extremely helpful!

**Comment by aeneasr** [comment]
We run an multi-region system in the Ory Network where this is solved with proprietary code because there are more things than just CockroachDB to make it work. The open source variant of Ory Kratos does not have these optimizations because they require things like global TCP routing that works for cockroach and more.

Please note that every software system will have this problem (Keycloak and others) because of physics. If your SQL queries travel 400ms through deep sea cables, then every SQL query will be slow. Unless the application is heavily optimized for global throughput against the specific RDBMS, you will always have high latencies. No software can solve this unless it's heavily optimized for Cockroach.

**Reply by vstreame** [reply]
@aeneasr I understand what you're saying, but the rest of our system has been tuned for that. I was under the mistaken idea that the way CockroachDB works is the same how DynamoDB Global Tables works where all data is replicated to every region by default and then becomes eventually consistent. But now know that's not the case. 

Would Kraots break if we started to change it's schema ourselves and start marking tables as [GLOBAL](https://www.cockroachlabs.com/docs/stable/global-tables.html)?

**Reply by aeneasr** [reply]
GLOBAL tables do not respect regional data privacy laws, such as GDPR. While technically possible (global tables are **not** a panacea to latency, they also have physical limitations - how fast light travels for example), global tables are violating the law in many regions of the world as soon as they contain personal information (Europe, Brazil, California, Singapore, the list goes on ...), which includes many tables in Ory. We solved this in the Ory Network but it's unfortunately not available as open source code / knowledge and only on our managed network infrastructure.
